go-client/client/client.go

```diff
@@ -0,0 +1,262 @@
+//
+// If this function completes successfully, the number of tables returned will always match the
+// number of query nodes passed.
+//
+// This may return a QueryError if the query is invalid.
+func (client *Client) ExecQuery(ctx context.Context, nodes ...QueryNode) ([]*TableHandle, error) {
+	if client.Closed() {
+		return nil, ErrClosedClient
+	}
+
+	return execQuery(client, ctx, nodes)
+}
+
+// Closes the connection to the server and frees any associated resources.
+// Once this method is called, the client and any TableHandles from it cannot be used.
+func (client *Client) Close() {
+	client.sessionStub.Close()
+	client.appStub.Close()
+	if client.grpcChannel != nil {
+		client.grpcChannel.Close()
+		client.grpcChannel = nil
+	}
+}
+
+// This is thread-safe
+func (client *Client) withToken(ctx context.Context) context.Context {
+	return metadata.NewOutgoingContext(context.Background(), metadata.Pairs("deephaven_session_id", string(client.getToken())))
+}
+
+// Directly uploads and executes a script on the deephaven server.
+// The script language depends on the scriptLanguage argument passed when creating the client.
+func (client *Client) RunScript(context context.Context, script string) error {
+	// This has to shadow the consoleStub method in order to handle the listfields loop
+
+	if client.Closed() {
+		return ErrClosedClient
+	}
+
+	restartLoop := client.appStub.isListing()
+	client.appStub.cancelListLoop()
+
+	if restartLoop {
+		// Clear out the table list to avoid any duplicate entries
+		client.tables = make(map[fieldId]*TableHandle)
+	}
+
+	err := client.consoleStub.RunScript(context, script)
+	if err != nil {
+		client.FetchTables(context, FetchRepeating) // At least try to restart this
+		return err
+	}
+
+	if restartLoop {
+		err = client.FetchTables(context, FetchRepeating)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// This is thread-safe
+func (client *Client) handleScriptChanges(resp *consolepb2.ExecuteCommandResponse) {
+	client.handleFieldChanges(resp.Changes)
+}
+
+// This is thread-safe
+func (client *Client) handleFieldChanges(resp *applicationpb2.FieldsChangeUpdate) {
+	client.tablesLock.Lock()
+	defer client.tablesLock.Unlock()
+
+	for _, created := range resp.Created {
+		if created.TypedTicket.Type == "Table" {
+			fieldId := fieldId{appId: created.ApplicationId, fieldName: created.FieldName}
+			client.tables[fieldId] = newTableHandle(client, created.TypedTicket.Ticket, nil, 0, false)
+		}
+	}
+
+	for _, updated := range resp.Updated {
+		if updated.TypedTicket.Type == "Table" {
+			fieldId := fieldId{appId: updated.ApplicationId, fieldName: updated.FieldName}
+			client.tables[fieldId] = newTableHandle(client, updated.TypedTicket.Ticket, nil, 0, false)
+		}
+	}
+
+	for _, removed := range resp.Removed {
+		if removed.TypedTicket.Type == "Table" {
+			fieldId := fieldId{appId: removed.ApplicationId, fieldName: removed.FieldName}
+			delete(client.tables, fieldId)
+		}
+	}
+}
```

Comment:
Overall, need to think very carefully about the concurrency model, since Go loves 1M threads.